token-lending/program/src/state/reserve.rs

```diff
@@ -953,6 +988,75 @@ mod test {
             }
         }
 
+        #[test]
+        fn minimum_collateral(
+            collateral_amount in 0..=u32::MAX as u64,
```

Comment:
Can you change this to a range over `borrow_amount` instead? That's more what's being tested

Reply:
Hmm. If `borrow_amount` is what we want to test, then this tests the same inputs as https://github.com/solana-labs/solana-program-library/blob/a99e052566e52967ce0404f2e77b3799835aeb54/token-lending/program/src/state/reserve.rs#L1014-L1019

Using `borrow_amount` with `required_collateral_for_borrow` was [an earlier approach](https://github.com/solana-labs/solana-program-library/pull/1229/files/f65021b4d09088be30792deceeaeebf6b46b1983#diff-724dc1de5d8ee8e4daf8d2a6d0c0c12bde2a03921e3cc74bcb75ab0caf3d00f1R1484-R1492) we considered. If this approach is okay, then maybe I can just remove this function and test?

Reply:
Required collateral for borrow has special logic for calculating the impact price when taking out a new borrow position. We don't want to do that for withdraws because the slippage could prevent borrowers from taking out a reasonable amount of their collateral. I think I was wrong here, the withdraw should allow borrowers to take out collateral just up until the liquidation point if they want to.

We could even add a parameter for them to decide what they want the resulting LTV to be after their withdrawal. For example, if liquidation is at 80% and the user wants to live on the edge, they can specify that they want to pull out up to 78% LTV. @bartosz-lipinski do you think allowing users to specify an LTV as a constraint would be useful?

Reply:
> Required collateral for borrow has special logic for calculating the impact price when taking out a new borrow position. We don't want to do that for withdraws because the slippage could prevent borrowers from taking out a reasonable amount of their collateral.

What happens if you max borrow (using the impact price) then immediately withdraw collateral (using the best price)? It seems that if your max borrow is determined by your collateral, then your min collateral should be determined by your borrow using the same price.

> the withdraw should allow borrowers to take out collateral just up until the liquidation point if they want to

Similar to above, but could you take out a loan, manipulate the best price for the collateral and/or liquidity to decrease the obligation's LTV, and withdraw more collateral than you should?

Reply:
Yeah, good points. Makes the most sense to keep borrow/withdraw collateral paths the same otherwise we just have a loophole. So, I think agree with your earlier point now about just using the same method and nixing minimum_collateral..

Reply:
Removed in dcdcf9e